slurk/views/chat/events.py

```diff
@@ -49,76 +49,92 @@ def typed_message(payload):
+            target = str(room.id)
+            private = False
+
+    user = dict(id=current_user.get_id(), name=current_user.name)
+
+    socketio.emit(
+        event,
+        dict(
+            user=user,
+            room=str(room.id) if room else None,
+            timestamp=str(datetime.utcnow()),
+            private=private,
+            **data,
+        ),
+        room=target,
+        broadcast=broadcast,
+    )
+
+    Log.add(
+        event=event,
+        user=current_user,
+        room=room,
+        receiver=receiver if private and not broadcast else None,
+        data=data,
+    )
+
+    if room:
+        for room in current_user.rooms:
+            socketio.emit("stop_typing", {"user": user}, room=str(room.id))
+
+    return True
+
+
 @socketio.event
 @login_required
 def text(payload):
     current_user_id = current_user.get_id()
     if not current_user_id:
         return False, "invalid session id"
-    if not current_user.token.permissions.send_message:
+
+    html = payload.get("html", False)
+    if not current_user.token.permissions.send_html_message and (
+        html or not current_user.token.permissions.send_message
+    ):
         return False, "insufficient rights"
     if "msg" not in payload:
         return False, 'missing argument: "msg"'
-    if "room" not in payload:
-        return False, 'missing argument: "room"'
-
-    broadcast = payload.get("broadcast", False)
-    if broadcast and not current_user.token.permissions.broadcast:
-        return False, "insufficient rights"
 
-    db = current_app.session
-    room = db.query(Room).get(payload["room"])
-    if not room:
-        return False, "Room not found"
-    if current_user not in room.users:
-        return False, "Not in room"
-
-    if room.layout.read_only:
-        return False, 'Room "%s" is read-only' % room.label
-
-    if "receiver_id" in payload:
-        if not current_user.token.permissions.private_message:
-            return False, "You are not allowed to send private text messages"
-        receiver_id = payload["receiver_id"]
-        receiver = db.query(User).get(receiver_id)
-        if not receiver or not receiver.session_id:
-            return False, 'User "%s" does not exist' % receiver_id
-        private = True
-    else:
-        receiver = None
-        private = False
-
-    user = {
-        "id": current_user_id,
-        "name": current_user.name,
-    }
-
-    socketio.emit(
+    return emit_message(
         "text_message",
-        {
-            "msg": payload["msg"],
-            "user": user,
-            "room": str(room.id) if room else None,
-            "timestamp": str(datetime.utcnow()),
-            "private": private,
-            "html": payload.get("html", False),
-        },
-        room=receiver.session_id if private else str(room.id),
-        broadcast=broadcast,
+        payload,
+        data=dict(message=payload["msg"], html=html),
```

Comment:
For the log the key was called `message`, but for the emit the key was called `msg`. The scripts `display_text` and `markdown` assume a key `msg`, so we quite likely need to adapt them.